crates/core/tedge_monitor/src/main.rs

```diff
@@ -0,0 +1,166 @@
+
+    loop {
+        if get_process_id(name, m_type) > 0 {
+            for event in connection.iter() {
+                match event {
+                    Ok(Event::Incoming(Packet::SubAck(_))) => {
+                        client.publish(req_topic, AtLeastOnce, false, "").unwrap();
+                    }
+                    Ok(Event::Incoming(Packet::PubAck(_))) => {
+                        continue;
+                        // The request has been sent
+                    }
+                    Ok(Event::Incoming(Packet::Publish(_response))) => {
+                        // Received response from thin-edge service, update the status to systemd on behalf of service
+                        let pid = get_process_id(name, m_type);
+                        notify_systemd(pid, "WATCHDOG=1")?;
+                        break;
+                    }
+                    Ok(Event::Outgoing(Outgoing::PingReq)) => {
+                        // No messages have been received for a while
+                        eprintln!("Local MQTT publish has timed out.");
+                        break;
+                    }
+                    Ok(Event::Incoming(Incoming::Disconnect)) => {
+                        eprintln!("ERROR: Disconnected");
+                        break;
+                    }
+                    Err(err) => {
+                        eprintln!("ERROR: {:?}", err);
+                        // client.subscribe(res_topic, AtLeastOnce).unwrap();
+                        break;
+                    }
+                    _ => {}
+                }
+            }
+        } else {
+            thread::sleep(time::Duration::from_secs(5));
+        }
+        thread::sleep(time::Duration::from_secs(2));
+        client.publish(req_topic, AtLeastOnce, false, "").unwrap();
+    }
+}
+
+fn get_mqtt_options(client_id: &str) -> MqttOptions {
+    let tedge_config_location =
+        tedge_config::TEdgeConfigLocation::from_custom_root(DEFAULT_TEDGE_CONFIG_PATH);
+    let config_repository = tedge_config::TEdgeConfigRepository::new(tedge_config_location.clone());
+    let tedge_config = config_repository.load().unwrap();
+    let host = tedge_config
+        .query(MqttBindAddressSetting)
+        .unwrap()
+        .to_string();
+    let port = tedge_config.query(MqttPortSetting).unwrap().into();
+    dbg!(&host);
+    dbg!(&port);
+    MqttOptions::new(client_id, host, port)
+}
+
+fn get_process_id(daemon_name: &str, mapper_type: Option<&str>) -> u32 {
+    let s = System::new_all();
+
+    for process in s.processes_by_exact_name(daemon_name) {
+        match mapper_type {
+            Some(m_type) => {
+                for subcmd in process.cmd().iter() {
+                    if m_type.eq(subcmd) {
+                        return process.pid().as_u32();
+                    }
+                }
+            }
+            None => {
+                return process.pid().as_u32();
+            }
+        }
+    }
+    return 0;
+}
```

Comment:
Can't the processes report their PID in the status msg?

Reply:
Yes, now that feature is implemented in the mapper and agent services to send the PID with the status response. I am using PID sent by these services.